File: mahaam-api-cs/Src/Feat/Plan/PlanController.cs
using System.Net.Mime;
using Mahaam.Infra;
using Microsoft.AspNetCore.Mvc;

namespace Mahaam.Feat.Plans;

public interface IPlanController
{
	Task<IActionResult> Create(PlanIn plan);
	Task<IActionResult> Update(PlanIn plan);
	Task<IActionResult> Delete(Guid id);
	Task<IActionResult> Share(Guid id, string email);
	Task<IActionResult> Unshare(Guid id, string email);
	Task<IActionResult> Leave(Guid id);
	Task<IActionResult> UpdateType(Guid id, string type);
	Task<IActionResult> ReOrder(string type, int oldOrder, int newOrder);
	Task<IActionResult> GetOne(Guid id);
	Task<IActionResult> GetMany(string? type);
}


[ApiController]
[Route("plans")]
public class PlanController(IPlanService planService) : ControllerBase, IPlanController
{

	[HttpPost]
	[Consumes(MediaTypeNames.Application.Json)]
	public async Task<IActionResult> Create([FromBody] PlanIn plan)
	{
		Rule.OneAtLeastRequired([plan.Title, plan.Starts, plan.Ends], "title or starts or ends is required");
		var id = await planService.Create(plan);
		return Created($"/plans/{id}", id);
	}

	[HttpPut]
	[Consumes(MediaTypeNames.Application.Json)]
	public async Task<IActionResult> Update([FromBody] PlanIn plan)
	{
		Rule.Required(plan.Id, "Id");
		Rule.OneAtLeastRequired([plan.Title, plan.Starts, plan.Ends], "title or starts or ends is required");
		await planService.Update(plan);
		return Ok();
	}

	[HttpDelete]
	[Route("{id}")]
	public async Task<IActionResult> Delete(Guid id)
	{
		Rule.Required(id, "id");
		await planService.Delete(id);
		return NoContent();
	}

	[HttpPatch]
	[Route("{id}/share")]
	public async Task<IActionResult> Share(Guid id, [FromForm] string email)
	{
		Rule.Required(id, "id");
		Rule.Required(email, "email");
		await planService.Share(id, email);
		return Ok();
	}

	[HttpPatch]
	[Route("{id}/unshare")]
	public async Task<IActionResult> Unshare(Guid id, [FromForm] string email)
	{
		Rule.Required(id, "id");
		Rule.Required(email, "email");
		await planService.Unshare(id, email);
		return Ok();
	}

	[HttpPatch]
	[Route("{id}/leave")]
	public async Task<IActionResult> Leave(Guid id)
	{
		Rule.Required(id, "id");
		await planService.Leave(id);
		return Ok();
	}

	[HttpPatch]
	[Route("{id}/type")]
	public async Task<IActionResult> UpdateType(Guid id, [FromForm] string type)
	{
		Rule.Required(id, "id");
		Rule.Required(type, "type");
		Rule.In(type, PlanType.All);
		await planService.UpdateType(id, type);
		return Ok();
	}

	[HttpPatch]
	[Route("reorder")]
	public async Task<IActionResult> ReOrder([FromForm] string type, [FromForm] int oldOrder, [FromForm] int newOrder)
	{
		Rule.Required(type, "type");
		Rule.In(type, PlanType.All);
		Rule.Required(oldOrder, "oldOrder");
		Rule.Required(newOrder, "newOrder");
		await planService.ReOrder(type, oldOrder, newOrder);
		return Ok();
	}

	[HttpGet]
	[Route("{id}")]
	public async Task<IActionResult> GetOne(Guid id)
	{
		Rule.Required(id, "id");

		var plan = await planService.GetOne(id);
		return Ok(plan);
	}

	[HttpGet]
	[Route("")]
	public async Task<IActionResult> GetMany([FromQuery] string? type)
	{
		if (type is not null) { Rule.In(type, PlanType.All); }
		else { type = PlanType.Main; }

		var plans = await planService.GetMany(type);
		return Ok(plans);
	}
}

File: mahaam-api-java/src/main/java/mahaam/feat/plan/PlanController.java
package mahaam.feat.plan;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.DELETE;
import jakarta.ws.rs.FormParam;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.PATCH;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.PUT;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.QueryParam;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import mahaam.feat.plan.PlanModel.Plan;
import mahaam.feat.plan.PlanModel.PlanIn;
import mahaam.feat.plan.PlanModel.PlanType;
import mahaam.infra.Json;
import mahaam.infra.Rule;

public interface PlanController {
	Response create(PlanIn plan);

	Response update(PlanIn plan);

	Response delete(UUID id);

	Response share(UUID id, String email);

	Response unshare(UUID id, String email);

	Response leave(UUID id);

	Response updateType(UUID id, String type);

	Response reOrder(String type, int oldOrder, int newOrder);

	Response getOne(UUID planId);

	Response getMany(String type);
}

@ApplicationScoped
@Path("/plans")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
class DefaultPlanController implements PlanController {

	@Inject
	PlanService planService;

	@POST
	public Response create(PlanIn plan) {
		Rule.required(plan, "plan");
		Rule.oneAtLeastRequired(Arrays.asList(plan.title, plan.starts, plan.ends),
				"title or starts or ends is required");

		UUID id = planService.create(plan);
		return Response.status(Response.Status.CREATED).entity(Json.toString(id)).build();
	}

	@PUT
	public Response update(PlanIn plan) {
		Rule.required(plan, "plan");
		Rule.required(plan.id, "Id");
		Rule.oneAtLeastRequired(Arrays.asList(plan.title, plan.starts, plan.ends),
				"title or starts or ends is required");

		planService.update(plan);
		return Response.ok().build();
	}

	@DELETE
	@Path("/{id}")
	public Response delete(@PathParam("id") UUID id) {
		Rule.required(id, "id");
		planService.delete(id);
		return Response.noContent().build();
	}

	@PATCH
	@Path("/{id}/share")
	@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
	public Response share(@PathParam("id") UUID id, @FormParam("email") String email) {
		Rule.required(id, "id");
		Rule.required(email, "email");

		planService.share(id, email);
		return Response.ok().build();
	}

	@PATCH
	@Path("/{id}/unshare")
	@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
	public Response unshare(@PathParam("id") UUID id, @FormParam("email") String email) {
		Rule.required(id, "id");
		Rule.required(email, "email");

		planService.unshare(id, email);
		return Response.ok().build();
	}

	@PATCH
	@Path("/{id}/leave")
	public Response leave(@PathParam("id") UUID id) {
		Rule.required(id, "id");
		planService.leave(id);
		return Response.ok().build();
	}

	@PATCH
	@Path("/{id}/type")
	@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
	public Response updateType(@PathParam("id") UUID id, @FormParam("type") String type) {
		Rule.required(id, "id");
		Rule.required(type, "type");
		Rule.in(type, PlanType.ALL);

		planService.updateType(id, type);
		return Response.ok().build();
	}

	@PATCH
	@Path("/reorder")
	@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
	public Response reOrder(@FormParam("type") String type, @FormParam("oldOrder") int oldOrder,
			@FormParam("newOrder") int newOrder) {
		Rule.required(type, "type");
		Rule.in(type, PlanType.ALL);
		Rule.required(oldOrder, "oldOrder");
		Rule.required(newOrder, "newOrder");
		planService.reOrder(type, oldOrder, newOrder);
		return Response.ok().build();
	}

	@GET
	@Path("/{planId}")
	public Response getOne(@PathParam("planId") UUID planId) {
		Rule.required(planId, "planId");

		Plan plan = planService.getOne(planId);
		return Response.ok().entity(Json.toString(plan)).build();
	}

	@GET
	public Response getMany(@QueryParam("type") String type) {
		if (type != null) {
			Rule.in(type, PlanType.ALL);
		} else {
			type = PlanType.MAIN;
		}

		List<Plan> plans = planService.getMany(type);
		return Response.ok().entity(Json.toString(plans)).build();
	}
}
